common/deadlock/deadlock.go

```diff
@@ -0,0 +1,222 @@
+		},
+		roots: roots,
+	}
+}
+
+func (dd *deadlockDetector) Start() error {
+	for _, root := range dd.roots {
+		loopCtx := &loopContext{
+			dd:   dd,
+			root: root,
+			ch:   make(chan common.PingCheck),
+		}
+		dd.loops.Go(loopCtx.run)
+	}
+	return nil
+}
+
+func (dd *deadlockDetector) Stop() error {
+	dd.loops.Cancel()
+	// don't wait for workers to exit, they may be blocked
+	return nil
+}
+
+func (dd *deadlockDetector) detected(name string) {
+	dd.logger.Error("deadlock detected", tag.Name(name))
+
+	if dd.config.FailHealthCheck() {
+		dd.logger.Info("marking grpc services unhealthy")
+		dd.healthServer.Shutdown()
+	}
+
+	if dd.config.DumpGoroutines() {
+		dd.dumpGoroutines()
+	}
+
+	if dd.config.AbortProcess() {
+		dd.logger.Fatal("deadlock detected", tag.Name(name))
+	}
+}
+
+func (dd *deadlockDetector) dumpGoroutines() {
+	profile := pprof.Lookup("goroutine")
+	if profile == nil {
+		dd.logger.Error("could not find goroutine profile")
+		return
+	}
+	var b strings.Builder
+	err := profile.WriteTo(&b, 1) // 1 is magic value that means "text format"
+	if err != nil {
+		dd.logger.Error("failed to get goroutine profile", tag.Error(err))
+		return
+	}
+	// write it as a single log line with embedded newlines.
+	// the value starts with "goroutine profile: total ...\n" so it should be clear
+	dd.logger.Info(b.String())
+}
+
+func (lc *loopContext) run(ctx context.Context) error {
+	lc.ctx = ctx
+	// start with one worker
+	go lc.worker()
+	atomic.StoreInt32(&lc.workers, 1)
+
+	for {
+		// ping blocks until it has passed all checks to a worker goroutine (using an
+		// unbuffered channel).
+		lc.ping([]common.Pingable{lc.root})
+
+		select {
+		case <-time.After(lc.dd.config.Interval()):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (lc *loopContext) ping(pingables []common.Pingable) {
+	for _, pingable := range pingables {
+		for _, check := range pingable.GetPingChecks() {
+			select {
+			case lc.ch <- check:
+			case <-lc.ctx.Done():
+			default:
+				// maybe add another worker if blocked
+				w := atomic.LoadInt32(&lc.workers)
+				if w < int32(lc.dd.config.MaxWorkersPerRoot()) && atomic.CompareAndSwapInt32(&lc.workers, w, w+1) {
+					go lc.worker()
+				}
+				// blocking send
+				select {
+				case lc.ch <- check:
+				case <-lc.ctx.Done():
+				}
+			}
+		}
+	}
+}
+
+func (lc *loopContext) worker() {
+	for check := range lc.ch {
+		lc.dd.logger.Debug("starting ping check", tag.Name(check.Name))
+
+		// Using AfterFunc is cheaper than creating another goroutine to be the waiter, since
+		// we expect to always cancel it. If the go runtime is so messed up that it can't
+		// create a goroutine, that's a bigger problem than we can handle.
+		t := time.AfterFunc(check.Timeout, func() {
+			if lc.ctx.Err() != nil {
+				// deadlock detector was stopped
+				return
+			}
+			lc.dd.detected(check.Name)
+		})
+		newPingables := check.Ping()
```

Comment:
we should emit some latency metric tagged with component name

Reply:
good idea. except... for shards, I gave each one a unique "name" so we can correlate it with other log lines. but I don't think we want one metric tag per shard. so I guess we need another field in PingCheck for metric scope?

hmm. should we try to subsume this stuff?
https://github.com/temporalio/temporal/blob/968c5765ff5bd7c3d8ab2c97e57bacacf5dde45a/service/history/shard/context_impl.go#L1503-L1519

that uses a timer on every single lock, this would be sampled infrequently. so less data, but it would be easier to add metrics on more locks.

Reply:
Yes, there is already one issue created asking to remove that per lock request metrics. I think the component name should not include shard id. The logger could write shard ID. Maybe it should have ComponentName() and Identity() where identity would contain shardID.

Reply:
Whichever is easier for you. We do plan to migrate all metrics to the new MetricsHandler, but that is a separate work stream.

Reply:
I did it the easy (old) way for now. Will need a little work to migrate.